[gltf] Add typings for glTF 2.0

donmccurdy opened the pull request:
Adds typings for glTF 2.0, a file format for 3D assets.

/cc @bghgary Could you confirm this still works as expected in BabylonJS? I made a couple (hopefully minor) changes:

1. Added typings for `bufferView.target`
2. Replaced all enums with a type and a namespace, like:

```typescript
/**
 * Interpolation algorithm
 */
export type AnimationSamplerInterpolation = "LINEAR" | "STEP" | "CUBICSPLINE";
export namespace AnimationSamplerInterpolation {
    /**
     * The animated values are linearly interpolated between keyframes
     */
    const LINEAR = "LINEAR";
    /**
     * The animated values remain constant to the output of the first keyframe, until the next keyframe
     */
    const STEP = "STEP";
    /**
     * The animation's interpolation is computed using a cubic spline with specified tangents
     */
    const CUBICSPLINE = "CUBICSPLINE";
}
```

The step above is redundant, but without that a glTF object does not validate correctly, because enum strings are opaque. See: https://github.com/Microsoft/TypeScript/issues/17690.

Also note the "definitions by" header — we'll automatically be pinged if someone opens a PR on these typings. Should anyone else be listed?

***

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

sheetalkamat commented on types/gltf/tslint.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "extends": "dtslint/dt.json",
+    "rules": {
```
When adding new definitions you cannot have any rules here. Only extends clause.

donmccurdy replied:
Hm I see. I take that to mean that `IFoo` naming is disallowed on all typings in this repository?

@bghgary are you OK with the renaming this would involve for BabylonJS?

bghgary replied:
This is the naming convention we are using in BabylonJS for many interfaces. Types can be aliased when imported (especially with modules), so hopefully it will work out.

donmccurdy replied:
I'll remove the `I` prefixes then, but let me know if there is any issue.

donmccurdy replied:
Done ✅

sheetalkamat commented on types/gltf/index.d.ts:
```diff
@@ -0,0 +1,890 @@
+ */
+export type TextureWrapMode = 33071 | 33648 | 10497;
+export namespace TextureWrapMode {
+    /**
```
You cannot have these constants here.
```typescript
var a = require("gltf");
console.log(typeof a.TextureWrapMode);
```
prints `undefined`

bghgary replied:
`const enum` !== `enum`. My understanding is that `const enum` are okay in type declarations, which is what the original file uses.

donmccurdy replied:
Ok thanks, I'll revert to `const enum` then.

donmccurdy replied:
It appears that `const enum` is not allowed here either: https://github.com/Microsoft/dtslint/blob/master/docs/no-const-enum.md. Is there another way to include enum values with typings? Or should we be publishing those in an actual package?

sheetalkamat commented on types/gltf/index.d.ts:
```diff
@@ -0,0 +1,890 @@
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.1
+
+export as namespace GLTF;
```
This should only export `collada2gltf` as string? Do not see anything else being added by the module.

bghgary replied:
I'm a bit confused by this comment. Where do you see `collada2gltf` in the typing?

sheetalkamat replied:
```js
var a = require("gltf");
console.log(typeof a);
console.log(typeof a.collada2gltf );
console.log(typeof a.TextureWrapMode);
```

Shows:
```
object
string
undefined
```

sheetalkamat replied:
That lib is declaring all types and nothing in value space. (no enums, const, classes, functions etc) Yours seem to declare more than that.

bghgary replied:
@donmccurdy I chatted with @sheetalkamat offline. Looks like the confusion here is that gltf is already a npm package and it has collada2gltf in it right now. We either pick a different name or fix the existing package.

bghgary replied:
Might also consider following https://docs.npmjs.com/misc/disputes to figure out what to do.

donmccurdy replied:
> Looks like the confusion here is that gltf is already a npm package and it has collada2gltf in it right now.

That is unfortunate. 😕 I do think we should try to have the existing package renamed, then, but it may be tricky because there is no `gltf` library to replace it, just a format.

bghgary replied:
I was thinking we can possibly put the markdown and schema files in it.

donmccurdy replied:
Ok, that seems like a good idea. I wonder if we should consider using scoped packages at some point. We do own the `@khronosgroup/*` namespace already. 🤔

bghgary commented:
> @bghgary Could you confirm this still works as expected in BabylonJS?

Yes, I can try later this week, but let's fix the issues first.

> Replaced all enums with a type and a namespace

I'm not sure this is the right thing to do. There is definitely some weirdness when trying to assign a string to a string enum, but you can do it easily by casting the string to the string enum type.

```
accessor.type = "SCALAR" as AccessorType
```

In practice, this normally isn't necessary. In BabylonJS, we have importers and exporters for glTF and we don't do this anywhere in the code.

> Should anyone else be listed?

I think this is fine for now. We can always add more later.

donmccurdy commented:
It appears that `const enum` is not allowed here either: https://github.com/Microsoft/dtslint/blob/master/docs/no-const-enum.md. Is there another way to include enum values with typings? Or should we be publishing those in an actual package?

Other than that I think I've addressed the feedback.

bghgary commented:
Looks like many people use it in typings but they put `// tslint:disable-next-line:no-const-enum`.

bghgary commented:
See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#lint

bghgary commented:
@sheetalkamat Yes, we are working on that.

sheetalkamat commented:
Why do you need enum? Why not type ? eg: 
```ts
type AccessorComponentTypeBYTE = 5120;
```

bghgary commented:
This typings came from [BabylonJS](https://github.com/BabylonJS/Babylon.js/blob/master/dist/babylon.glTF2Interface.d.ts). I was hoping to replace the one in BabylonJS with this one. Using constants will make it no longer backwards compatible.

Enums are also a bit easy to use than a simple value in typescript especially with code completion, though I understand it doesn't work in some cases.

donmccurdy commented:
Added `tslint:disable-next-line`, tests are passing again. I guess the alternative would be publishing the enum values in a separate package? I would prefer enums to simple values.

@bghgary should we do anything to get ahead of versioning on this? Considering renaming the namespace as `GLTF_v2_0` or similar.

donmccurdy commented:
> As discussed this cant be in gltf since there exist npm package with that name

Would there be an issue if there were _no_ npm package with that name? Or simply a stub package that we manage? Similar to `geojson`, these are typings for a JSON-based data type, not a specific package.

bghgary commented:
> should we do anything to get ahead of versioning on this? Considering renaming the namespace as GLTF_v2_0 or similar.

In BabylonJS, I used GLTF2 as the namespace for these interfaces. Dot revisions are expected to be forward and backward compatible, so I don't think you need them in the namespace.

I think my preference would be to call the module GLTF2 and don't use namespaces at all. The import statement can choose the name to put the interfaces in and most likely people will just call it the same as the module. This also avoids the conflict with the existing glTF module.

donmccurdy commented:
Learning new things about TypeScript here. 😅

Trying to copy from the BabylonJS example, I get:

> The internal 'module' syntax is deprecated, use the 'namespace' keyword instead.

Converting that to an external module:

> File has only 1 module declaration — write it as an external module. See:https://github.com/Microsoft/dtslint/blob/master/docs/no-single-declare-module.md

So I think we're back to namespaces? I've renamed it to GLTF2, and made it a non-default export so GLTF3+ can be included in the future.

bghgary commented:
I unfortunately don't know that much about modules/namespaces either. We are the process of switching to modules in BabylonJS. Once that is done, we will be able to test this out, hopefully some time next week.

DanielRosenwasser commented:
There's already a `gltf` package on npm which seems potentially unrelated (since it has an export called `collada2gltf` and I see nothing of the sort here).

donmccurdy commented:
Note that like the existing `geojson` typings, glTF is a JSON-based file format — there is no one official glTF JS library. These typings allow TypeScript code to typecheck with glTF content. For example:

```typescript
const model = JSON.parse( content ) as GLTF2.GLTF;
console.log( model.asset.generator );
```

The existing `gltf` package is a converter from another format (COLLADA) into glTF. COLLADA is only one of many formats that could be converted to glTF, so its use of the name `gltf` is unfortunate. We're planning to resolve that, by renaming the existing package and (a) replacing it with a new stub package, or (b) leaving nothing in its place.

Is this reasonable, or should we be taking another approach here?

DanielRosenwasser commented:
> there is no one official glTF JS library

Sorry I should have provided more details in my last response. There is already a `gltf` packageon npm. Providing this package to DefinitelyTyped means we'll publish an `@types/gltf` npm package which wouldn't reflect the `gltf` package - but ideally `@types` packages should always reflect npm package names when possible.

On the other hand, the `gltf` package gets <70 downloads a month, so maybe it's not worth it. It's really up to you.

donmccurdy commented:
@DanielRosenwasser thanks, I'm aware of the `gltf` NPM package. If necessary, we can contact the author and NPM to have it renamed. But the key issue is that the typings I'm proposing in this PR are not, and will never be, really associated with any NPM package. When you say...

> ideally @types packages should always reflect npm package names when possible.

...what is the ideal outcome in this case? Options, as I see them:

- [ ] The existing `gltf` package remains as-is. It's lower-profile than the format itself, and (hopefully) won't confuse many users.
- [ ] We ask the author and NPM to rename the package. Now there is no `gltf` package to confuse with these typings, but I don't know if there's anything to stop another user from publishing a new `gltf` package.
- [ ] We ask the author and NPM to rename the package, and then publish a new placeholder glTF package in its place. That package would probably not do much of anything.

> We are the process of switching to modules in BabylonJS. Once that is done, we will be able to test this out, hopefully some time next week.

@bghgary thanks, any updates on this?

bghgary commented:
> the key issue is that the typings I'm proposing in this PR are not, and will never be, really associated with any NPM package

bghgary commented:
> the key issue is that the typings I'm proposing in this PR are not, and will never be, really associated with any NPM package

I'm wondering if we should be creating a DefinitelyTyped entry at all. Perhaps, this should just be a normal npm package?

>> We are the process of switching to modules in BabylonJS. Once that is done, we will be able to test this out, hopefully some time next week.
>
> @bghgary thanks, any updates on this?

This is taking much longer than expected, so unfortunately, we can't test these change out right now.

donmccurdy commented:
@bghgary Yeah, similarly I'm coming to the conclusion that we should just publish a normal NPM module containing the constants (useable in both JS and TS) and typings (only useable in TS) for the full specification. I don't know of any other examples, but it seems doable. We could publish that as `gltf` or `@khronosgroup/gltf`.

orta commented:
I'd recommend shipping the types as a normal NPM module also, I'm going to close this PR (as it's been about a year since the last comment) 

If you would strongly prefer it on definitely typed, I'd recommend creating an npm module which you own `gltf-format` then making the PR here with these types to associate with that module. That's probably more effort than owning the types in your own module though.
